### DISCUSSION TITLE
routeChange Event Listener for Next.js

### DISCUSSION THREAD

**Discussion by 0x15BA88FF** [vercel/next.js#37638] [Help] [answered]
iv been having a problem with listening to route changes in Next.js since it is Single Paged Web App, I've tried tried a work around to check the URL every second and update the page accordingly but it's probably not the most efficient thing to do.
```
useEffect(() => {
    setInterval(() => {   
        updateRoute()
    }, 1000);
}, [])
```

just reaching out to see if there are another work around or if an ```onRouteChange(() => {})``` event listener can be introduced into Next.js

**Comment by icyJoseph** [comment, marked as answer]
Use either, in `_app` most preferably:

### With the Router singleton instance

```js
import Router from 'next/router'

// this can be outside of any component
Router.events.on('routeChangeComplete', (url, options) => console.log({url, options});

// rest of file...
function MyApp() {}
```

## The React hooks way (also preferred)

```js
import {useRouter} from 'next/router';

// inside a React function component

const router = useRouter();

useEffect(() => {
  const handler = (url, options) => console.log({url, options});
  router.events.on('routeChangeComplete', handler);

  return () => router.events.off('routeChangeComplete', handler);
}, [])
```

Please refer to this page https://nextjs.org/docs/api-reference/next/router#routerevents, from the documentation, for more information on available events.

**Reply by leerob** [reply]
Thank you!

**Reply by 0x15BA88FF** [reply]
Thanks
